src/client/binary_client.cpp

```diff
@@ -260,11 +260,40 @@ namespace
       return response;
     }
 
-    virtual ActivateSessionResponse ActivateSession()
+    ActivateSessionResponse ActivateSession(const SessionData &session_data) override
     {
       if (Debug)  { std::cout << "binary_client| ActivateSession -->" << std::endl; }
       ActivateSessionRequest request;
       request.Parameters.LocaleIDs.push_back("en");
+
+      Common::Uri uri(Params.EndpointUrl);
+      std::string user = uri.User();
+      std::string password = uri.Password();
+      bool user_identify_token_found = false;
+      for(auto ep : session_data.ServerEndpoints) {
+        if(ep.SecurityMode == MSM_NONE) {
+          for(auto token : ep.UserIdentifyTokens) {
+            if(user.empty()) {
+              if(token.TokenType == UserIdentifyTokenType::ANONYMOUS) {
+                request.Parameters.IdentifyToken.setPolicyID(token.PolicyID);
+                user_identify_token_found = true;
+                break;
+              }
+            }
+            else {
+              if(token.TokenType == UserIdentifyTokenType::USERNAME) {
+                request.Parameters.IdentifyToken.setPolicyID(token.PolicyID);
+                request.Parameters.IdentifyToken.setUser(user, password);
+                user_identify_token_found = true;
+                break;
+              }
+            }
+          }
+        }
+      }
+      if(!user_identify_token_found) {
+        throw std::runtime_error("Cannot find suitable user identify token for session");
+      }
```

Comment:
The correct think is to pass UpdateSessionParamaters to ActivateSession method. And remove all guessing from that method, The idea with BinaryClient is to offer a raw opc-ua interface, then move all guessing stuff to client.cpp which is the high level class.